### PR TITLE
Feature/domino head for dflash

### LINKED ISSUE #46707: [RFC]: Support for Domino speculative decoding in vLLM
# [RFC]: Support for Domino speculative decoding in vLLM

> **Provenance.** Technical claims below are grounded in: the Domino paper
> (arXiv 2605.29707v1); the public `Huang2020/Qwen3-8B-Domino-b16` checkpoint
> (its `config.json` and the full `model.safetensors` tensor manifest, read
> directly); and a read of vLLM `main`'s `vllm/v1/spec_decode/` verified at
> commit `2396d91` (2026-06-25). The subsystem refactors often — re-confirm the
> vLLM line references against current main before relying on them.

## Motivation

vLLM supports a family of model-based speculative-decoding methods (EAGLE /
EAGLE-3, MTP, PARD, DFlash, draft-model). **Domino is not currently available
in vLLM.** Public Domino draft checkpoints exist (e.g.
`Huang2020/Qwen3-8B-Domino-b16`) and the authors provide a reference
implementation, but it is, per the paper, "mainly adapted to SGLang" — there is
no path to serve Domino under vLLM today.

The core idea (Domino, Huang et al., arXiv 2605.29707): speculative decoding is
bottlenecked by a quality–cost tradeoff. Autoregressive drafters (EAGLE) model
causal dependencies between draft tokens but pay a sequential per-token cost;
parallel drafters (DFlash) draft a whole block in one forward pass but lose
intra-block causal structure. Domino keeps drafting parallel — a parallel
backbone produces preliminary block logits in one pass — then applies a
lightweight **Domino head** that injects causal information as a low-rank
**residual correction in logit space**, avoiding any repeated draft-model or
full-LM-head execution.

**Why it's worth adding (authors' results, on A100 — not yet reproduced under
vLLM):**
- Up to 5.49× end-to-end speedup (Transformers backend) and up to ~5.8×
  throughput (SGLang) on Qwen3.
- vs. DFlash specifically: +56M params (+5.3%), only +2.8% draft-then-verify
  latency, but +16.6% average acceptance length and +12.3% end-to-end speedup.
- Head ablation (same model, correction off→on): avg acceptance length
  3.49 → 4.19, avg speedup 2.84× → 3.31×.

These are the authors' numbers under their stack. Establishing what holds under
vLLM is precisely the goal of this work (see Benchmarking).

## Key framing: Domino is a head on top of DFlash, which vLLM already has

This is the crux of why the integration is tractable. The published Domino
backbone **is** the DFlash architecture (Chen et al., arXiv 2602.06036), and
vLLM already ships a `DFlashProposer` (`vllm/v1/spec_decode/dflash.py`). The
public checkpoint confirms this directly: its `architectures` is
`["DFlashDraftModel"]` and the Domino head is declared as
`dflash_config.projector_type = "domino"` — i.e. the checkpoint is a DFlash
draft model *plus* a Domino projector, not a new architecture.

So this RFC does **not** propose a drafter from scratch. It proposes adding the
~56M-parameter Domino correction head (a GRU + a low-rank MLP) on top of vLLM's
existing DFlash proposer path.

## Where this fits in vLLM (proposed design)

vLLM V1 spec-decode is a proposer/verifier architecture. The verifier (target
model + rejection sampler) is method-agnostic and is **not** modified here. All
method-specific logic lives in a proposer subclass of `SpecDecodeBaseProposer`
(`vllm/v1/spec_decode/llm_base_proposer.py`, class at line 62 as of `2396d91`).
`DFlashProposer` is such a subclass today.

Domino is structurally **a sequential correction on a parallel drafter**:
- DFlash is a *parallel* drafter — it drafts all K block positions in one
  forward pass.
- The base `propose()` takes a single-pass early-return when `parallel_drafting`
  is set (`if self.num_speculative_tokens == 1 or self.parallel_drafting:` in
  `llm_base_proposer.py`; the later `only_one_forward_pass = is_graph_capturing
  or self.parallel_drafting` path uses the same flag). So a parallel drafter
  makes exactly **one** draft-sampling call holding all K positions' base logits
  at once.
- Domino's correction is in *logit space* on those base logits (the paper is
  explicit: hidden-space correction would force re-running the full LM head per
  step). This is what makes a single sampling override the natural home for it.

### Proposed implementation: `DominoProposer` (mirrors `DFlashProposer`)

1. **`__init__`** — assert `method == "domino"`; `super().__init__(...,
   pass_hidden_states_to_model=True)` (the same flag `DFlashProposer` passes);
   read `pure_draft_prefix_len` (=1 in the public ckpt) and `gru_hidden_dim`
   (=1024) from `dflash_config`; expect `prefix_gru` / `embed_proj` on the draft
   model.
2. **`set_inputs_first_pass`** — call `super()` to inherit DFlash's
   cross-attention input layout unchanged, then stash the bonus tokens
   (`next_token_ids`). Target hidden states (`z`/`H_i`) are already stashed by
   the DFlash parent as `self._dflash_hidden_states` (`dflash.py`,
   `set_inputs_first_pass`). This is the same stash-on-self pattern DFlash
   already uses.
3. **Sampling override (the Domino head).** Because of the parallel single-pass
   layout, the sampling step receives all K positions' logits at once. Per the
   `_sample_from_logits` contract in `llm_base_proposer.py`, parallel drafters
   pass `batch_size * K` rows, **request-major** (K consecutive slots per
   request). Reshape to `[N, K, V]` and run the correction loop, which is a
   near-direct port of the reference's legacy path:
     - positions `< pure_draft_prefix_len`: argmax base logits, no correction.
     - prime the GRU on `[bonus, prefix_drafts]` embeddings.
     - for `i in range(prefix_len, K)`:
         `Δ_i = embed_proj(cat([H_i, S_{i-1}]))`; `logits_i = base_i + Δ_i`;
         `tok_i = argmax(logits_i)`; embed `tok_i`; advance GRU to `S_i`.
     - apply a draft→target token remap (`d2t`) only if the config defines one.
       The public `b16` checkpoint does **not** — its head emits full target
       vocab (see below), so no remap applies.

   **Design choice — see Alternatives.** The GRU state must thread across the K
   positions; the question is *where* that loop lives.

### New draft-model class: `DominoForCausalLM` (or extend the DFlash model class)

Carries the Domino head and loads its weights. **Verified against the public
`Qwen3-8B-Domino-b16` checkpoint** — keys and shapes read directly from the
`model.safetensors` manifest (62 tensors total):

| weight key | shape | role |
|---|---|---|
| `prefix_gru.weight_ih_l0` | `[3072, 4096]` | GRU input→gates; in=hidden_size 4096, 3072 = 3×1024 (r/z/n gates) |
| `prefix_gru.weight_hh_l0` | `[3072, 1024]` | GRU hidden→gates; gru_hidden_dim 1024 |
| `embed_proj.0.weight` | `[256, 5120]` | W1; in = **5120 = 4096 + 1024** = [H_i ; S_{i-1}] concat |
| `embed_proj.2.weight` | `[151936, 256]` | W2; low-rank r=256 → **full vocab** 151936 |

Notes that matter for the loader (all confirmed against the full manifest):
- **The entire head is bias-free.** The full state dict contains **no `bias`
  tensors anywhere** — neither GRU (`bias_ih_l0`/`bias_hh_l0` absent) nor the
  `embed_proj` Linears. Construct as `nn.GRU(4096, 1024, bias=False)` and both
  `embed_proj` Linears as `bias=False`, or weight-loading will look for bias
  tensors that don't exist.
- `embed_proj` is `Linear(5120→256, bias=False) → SiLU → Linear(256→151936,
  bias=False)`; the index `.0` / `.2` is an `nn.Sequential` with SiLU at `.1`.
- The correction emits a **full-vocab** logit bias, added to base logits before
  argmax — not a reduced-vocab correction. There is no `d2t`/`t2d` tensor in the
  checkpoint, consistent with `vocab_size == 151936` end to end.
- Backbone pieces (`fc.weight [4096, 20480]` = projection over the 5 captured
  target layers, 5×4096; `hidden_norm`, the 5 transformer `layers.{0..4}`,
  `norm`) are **inherited DFlash machinery**, already handled by `DFlashProposer`
  — not part of the Domino delta. They reinforce the "head on existing backbone"
  framing.

### Config / registration contract

From the public checkpoint's `config.json` (all values read directly):
- `architectures: ["DFlashDraftModel"]`, `model_type: "qwen3"`,
  `auto_map.AutoModel: "dflash.DFlashDraftModel"` (the HF path expects DFlash's
  custom modeling file — vLLM-native support is the actual work, vs.
  trust-remote-code).
- `dflash_config`: `projector_type="domino"`, `gru_hidden_dim=1024`,
  `emb_dim`/r `=256`, `pure_draft_prefix_len=1`, `mask_token_id=151669`,
  `shift_label=true`, `target_layer_ids=[1,9,17,25,33]`.
- Top level: `block_size=16`, `num_hidden_layers=5`, `hidden_size=4096`,
  `vocab_size=151936`.

Proposed: register `method="domino"` routing to `DominoProposer`; the draft
config must set `parallel_drafting` and carry `mask_token_id` **inside
`dflash_config`** — vLLM's `_init_parallel_drafting_params` reads
`dflash_config["mask_token_id"]` for this code path and raises if it is absent
(`llm_base_proposer.py`, lines ~331–349 at `2396d91`).

## Alternatives considered

**Alternative A — override the K-step loop in `propose()` to thread GRU state.**
Faithful to the reference (which owns its whole loop), but duplicates a large
part of the base class's biggest method and increases drift surface against
future spec-decode refactors.

**Proposed (B) — keep DFlash's single-pass layout; run the sequential GRU
correction inside an overridden sampling step.** Smaller, more stable override
surface; matches both the reference's "produce all K tokens in one call" shape
and the paper's logit-space design. The GRU is still walked sequentially over
the K rows, but *inside* that one sampling call. The control-flow precondition
(a single sampling call for parallel drafters) holds on current main — see the
`parallel_drafting` early-return above.

I propose **B**. One genuinely open design question for the team: should
`DominoProposer` subclass `DFlashProposer` directly, share a new intermediate
base, or stand alone? That depends on where the spec-decode subsystem is headed,
so I'd rather have maintainers' read than presume. (See Open Questions.)

## Proposed plan / phases (discrete reviewed PRs, not one drop)

1. `DominoForCausalLM` draft model + weight loading (the 4 head tensors above,
   all `bias=False`), reusing the DFlash backbone loader.
2. `DominoProposer` + `method="domino"` registration.
3. Correctness: token-parity vs the reference's **legacy** (cuDNN-GRU +
   nn.Sequential, cat-then-Linear) path. Methodology: greedy decoding, fixed
   seed, identical prompts, compare argmax token IDs per block position between
   vLLM `DominoProposer` and the reference. Target is exact per-position token
   parity on the draft block; the acceptance-length/speedup numbers are reported
   separately under Benchmarking and left blank until measured.
4. Benchmarking + docs.

**Benchmarking caveat (state plainly):** the paper's speedups are on A100 under
SGLang/Transformers; the reference is "mainly adapted to SGLang." Initial vLLM
validation here is on a T4 and is correctness-oriented — T4 perf is **not**
representative of Domino's intended hardware, and the efficiency-engineering the
authors use (fused Triton kernels, CUDA Graphs) is out of scope for v1. Do not
claim the paper's speedups will reproduce under vLLM; measuring that is the goal.

## Open questions

- **Subclassing (above):** `DominoProposer` subclasses `DFlashProposer`, a new
  shared base, or stands alone?
- **Head-offset special-casing.** In `llm_base_proposer.py` the per-request head
  offset is computed as
  `1 if (pass_hidden_states_to_model and self.method != "dflash") else 0`.
  DFlash is explicitly special-cased to `0`; a new `method="domino"` would take
  the `!= "dflash"` branch and get `1`. Whether Domino should join DFlash's
  special case (or whether the offset is correct as-is for it) needs a
  maintainer's eye — it affects input layout, not just registration.
- **Draft-config schema** vLLM expects for a *new* parallel-drafting method
  beyond `mask_token_id` (e.g. how `block_size`/`num_speculative_tokens` are
  surfaced for Domino).
- **Tree vs block:** Domino/DFlash are block-parallel (no draft tree); confirm
  the proposer stays on vLLM's block verification path, not tree decoding.

*Resolved during drafting (recorded so reviewers don't re-ask): the `embed_proj`
Linears carry no bias (whole head is bias-free), and this checkpoint has no
`d2t` remap (full-vocab head). Both verified against the `model.safetensors`
manifest.*

### PR DESCRIPTION
**Pull request by Eros483**
Closes #46707 (RFC). Based on training-side work in [vllm-project/speculators#685](https://github.com/vllm-project/speculators/pull/685).

## Purpose
Adds the inference-side changes needed for the Domino projector in DFlash,
enabling the full performance boost from Domino-trained draft models.

Training side experiments will be run to determine if this provides meaningful benefit.

Domino (arXiv 2605.29707) is a lightweight causal correction head (GRU +
low-rank MLP) that refines the parallel DFlash base logits using causal state
from previous draft tokens, operating entirely in logit space without
requiring additional forward passes.

Domino is configured as a `projector_type` sub-mode of DFlash
(`dflash_config.projector_type = "domino"`), not a separate method. This PR
adds:
- `DominoHead` module (`prefix_gru` + `embed_proj`) in `qwen3_dflash.py`
- Weight loading for Domino head tensors (bias-free, full-vocab correction)
- `_generate_domino_draft` method in `DFlashSpeculator` with GRU state
  threading across draft positions
- Routing in `_generate_draft` based on `projector_type`
- Documentation for DFlash/Domino usage
Original work by swylyu001 (<swylyu001@gmail.com>), rebased onto latest
`upstream/main` with conflict resolution against intervening DSpark and
Laguna XS 2.1 changes.

## Test Plan

- DFlash lookahead scheduler tests (CPU)
.venv/bin/python -m pytest tests/v1/spec_decode/test_dflash_lookahead.py -v

- DFlash proposer integration test
.venv/bin/python -m pytest tests/v1/spec_decode/test_eagle.py::test_set_inputs_first_pass_dflash -v

- Speculators correctness (requires ≥40GB GPU)
.venv/bin/python -m pytest tests/v1/spec_decode/test_speculators_correctness.py -v

## Test Results
- tests/v1/spec_decode/test_dflash_lookahead.py ........ 3/3 PASSED
- tests/v1/spec_decode/test_eagle.py::test_set_inputs_first_pass_dflash ... PASSED
- tests/v1/spec_decode/test_speculators_correctness.py ... 4 FAILED / 4 SKIPPED

The 4 failures in `test_speculators_correctness` are all `CUDA OOM` on an 8 GB
GPU — the Qwen3-8B target model requires ≥40 GB. Failures affect dflash, peagle,
and eagle3 tests identically, confirming no regression from this change.

## Future work
This work is currently focused simply on correctness, in an effort to match existing implementations from SGLang. If this implementation does backup the gains reported from Domino in the paper, it needs speed up from either torch compilation, triton or graph captures.

## Checklist
- [x] Purpose filled in (links RFC #46707 and speculators#685)
- [x] Test plan + results provided above
- [x] Documentation updated (dflash.md + README)
- [x] Linting passes (ruff, mypy, pre-commit)

---
AI assistance was used for this PR.
Co-authored-by: swylyu001 <swylyu001@gmail.com>
Signed-off-by: Eros483 <arnabmandal2912@gmail.com>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by orestis-z** [comment]
@Eros483 do you mind closing this in favor of https://github.com/vllm-project/vllm/pull/48241? alternatively do you mind merging in those changes if you agree? Then I would close https://github.com/vllm-project/vllm/pull/48241. In both scenarios https://github.com/vllm-project/vllm/pull/48241 is build on top of your changes so your contribution is preserved. Trying to avoid having multiple PRs open
